src/water_quality/mapping/algorithms.py

```diff
@@ -12,73 +13,315 @@
+        red_band, l_red = inst_bands["red"]
+        nir_band, l_nir = inst_bands["nir"]
+        swir_band, l_swir = inst_bands["swir"]
+
+        # Compute FAI
+        # FAI = Observed_NIR - Interpolated_NIR
+        # Observerd NIR is the Rayleigh-corrected reflectance in the NIR
+        # Interpolated_NIR = Red + (SWIR - Red) * (NIR_wl - Red_wl) / (SWIR_wl - Red_wl)
+        # Interpolated_NIR is a linear baseline formed by the red and SWIR bands.
+        # It follows the convention Slope-intercept form:
+        # y = mx+b, where m is the slope of the line and b is the y-intercept.
+
+        # Calculate the slope coefficient (m) for the baseline
+        # m = (NIR_wl - Red_wl) / (SWIR_wl - Red_wl)
+        m = (l_nir - l_red) / (l_swir - l_red)
+
+        # Calculate the baseline reflectance at NIR wavelength
+        interpolated_nir = ds[red_band] + (ds[swir_band] - ds[red_band]) * m
+        # Scale by 10000 assuming input reflectance is in 0-10000 range
+        fai = ds[nir_band] - interpolated_nir
+        fai.name = "FAI"
+        return fai / 10000
+
+
+def geomedian_FAI(ds: xr.Dataset) -> xr.Dataset:
+    """
+    Calculate the FAI (Floating Algae Index) across multiple instruments
+    and produce a combined weighted mean FAI for water pixels.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Input dataset containing the spectral bands.
+
+    Returns
+    -------
+    xr.DataArray
+        The updated input xarray Dataset with new bands:
+        - '{instrument}_fai' for each processed instrument (masked).
+        - 'agm_fai' (the final weighted average FAI, masked).
+    """
+
+    reference_mean = {
+        "fai": {"msi_agm": 0.0970, "oli_agm": 0.1015, "tm_agm": 0.0962},
+    }
+    # threshold = {
+    #     'fai' : {'msi_agm': 0.05   , 'oli_agm' : 0.05   , 'tm_agm': 0.05},
+    #    }
+    fai_threshold = 0.05
+    # Keep this order for consistent processing.
+    geomedian_instruments = ["msi_agm", "oli_agm", "tm_agm"]
+    # The instrument to use as the scaling reference
+    reference_inst = "msi_agm"
+
+    # Initialization for Weighted Average
+    mean_fai_weighted_sum = None
+    agm_count_total = None
+
+    for inst_agm in geomedian_instruments:
+        # Use the smad band as an indicator that data for the geomedian
+        # instrument exists in the dataset.
+        smad_band = f"{inst_agm}_smad"
+        if smad_band in ds.data_vars:
+            count_band = f"{inst_agm}_count"
+
+            # Scale factor based on the reference mean of the current instrument
+            # relative to the reference mean of the reference instrument.
+            scale = (
+                reference_mean["fai"][reference_inst]
+                / reference_mean["fai"][inst_agm]
+            )
+            # Calculate the FAI for the instrument and scale
+            fai_data = FAI(ds, inst_agm) * scale
+            # Replace all NaN values with 0s in FAI
+            fai_data = fai_data.fillna(0)
```

Comment:
not necessary because later filtering and masking steps will turn them back to nan?